### PR TITLE
Service implementation of authorised keys.

### PR DESCRIPTION
**Pull request by tlm**
This commit introduces the services layer for authorised keys. The logic implemented in the services layer is taken from what we already have in the key manager facade.

As an implementation detail we still expect authorised keys to live in model config for the forseeable future.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

There is no QA to be performed as part of this PR but I would appreciate a sanity check of the current logic from https://github.com/juju/juju/blob/main/apiserver/facades/client/keymanager/keymanager.go

## Documentation changes

Yes we will have to update https://juju.is/docs/juju/juju-remove-ssh-key

Previously in Juju versions less then 4.0 we only worked with fingerprints calculated using MD5 sums. This Pr introduces the change to use SHA256 sums. We should document this expectation to the user.

@tmihoc will need to work out how we coordinate this with the release.

## Links

**Jira card:** JUJU-5342

### CODE REVIEW

**Comment by tmihoc** [comment]
@tlm Please create a JIRA card for the documentation task. It should be under the same epic as the epic where the development work has happened. We can take it from there.

**Review by SimonRichardson — APPROVED** [review]
Change the wording the doc.go, otherwise it's good to go.

File: domain/keymanager/service/service.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+	// keyImporter is the [PublicKeyImporter] to use for fetching a users
+	// public key's for subject.
+	keyImporter PublicKeyImporter
+
+	// st is the provides the state access layer to this service.
+	st State
+}
+
+// State provides the access layer the [Service] needs for persisting and
+// retrieving a users public keys on a model.
```
Comment (manadart) [inline domain/keymanager/service/service.go:48]: "user's"

File: domain/keymanager/service/service.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+	// AddPublicKeyForUserIfNotFound will attempt to add the given set of public
+	// keys to the user. If the user already contains the public key it will be
+	// skipped and no [keyserrors.PublicKeyAlreadyExists] error will be returned.
+	AddPublicKeyForUserIfNotFound(context.Context, user.UUID, []keymanager.PublicKey) error
+
+	// GetPublicKeysForUser is responsible for returning all of the
+	// public keys for the current user in this model.
+	GetPublicKeysForUser(context.Context, user.UUID) ([]string, error)
+
+	// DeletePublicKeysForUser is responsible for removing the keys form the
```
Comment (manadart) [inline domain/keymanager/service/service.go:64]: s/form/from

File: domain/keymanager/service/service.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+			Fingerprint: parsedKey.Fingerprint(),
+			Key:         keyToAdd,
+		})
+	}
+
+	return s.st.AddPublicKeysForUser(ctx, userID, toAdd)
+}
+
+// DeletePublicKeysForUser removes the keys associated with targets from the
+// users list of public keys. Targets can be an arbitrary list of a
```
Comment (manadart) [inline domain/keymanager/service/service.go:142]: "user's"

File: domain/keymanager/types.go [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package keymanager
+
+type PublicKey struct {
```
Comment (manadart) [inline domain/keymanager/types.go:6]: Comments.

File: internal/ssh/importer/github.go [hunk 5]
```diff
@@ -48,7 +50,8 @@ const (
 // Github subject in this case a user and returning all of the public ssh keys
 // the user has for their profile.
 // The following errors can be expected:
-// - [SubjectNotFound] when the subject being asked for does not exist in
+// - [importererrors.SubjectNotFound] when the subject being asked for does not
+// exist in
```
Comment (manadart) [inline internal/ssh/importer/github.go:54]: "exist in" what?

File: domain/keymanager/service/service.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+) error {
+	if err := userID.Validate(); err != nil {
+		return fmt.Errorf(
+			"validating user id %q when importing public keys from %q: %w",
+			userID, subject.String(), err,
+		)
+	}
+
+	keys, err := s.keyImporter.FetchPublicKeysForSubject(ctx, subject)
+	if errors.Is(err, importererrors.NoResolver) {
```
Comment (manadart) [inline domain/keymanager/service/service.go:192]: I like switches for this kind of thing. No biggie.

**Comment by hpidcock** [comment]
Please use conventional commits.

File: domain/keymanager/service/service.go (hunk 6 above)
Comment (tlm) [reply domain/keymanager/service/service.go:192]: Done & Done

File: internal/ssh/importer/github.go (hunk 5 above)
Comment (tlm) [reply internal/ssh/importer/github.go:54]: Done & Done

File: domain/keymanager/types.go (hunk 4 above)
Comment (tlm) [reply domain/keymanager/types.go:6]: Done & Done

File: domain/keymanager/service/service.go (hunk 3 above)
Comment (tlm) [reply domain/keymanager/service/service.go:142]: Done & Done

File: domain/keymanager/service/service.go (hunk 2 above)
Comment (tlm) [reply domain/keymanager/service/service.go:64]: Done & Done

File: domain/keymanager/service/service.go (hunk 1 above)
Comment (tlm) [reply domain/keymanager/service/service.go:48]: Done & Done

File: domain/keymanager/service/service.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+// Service provides the means for interacting with a users underlying
+// public keys for a model.
+type Service struct {
+	// keyImporter is the [PublicKeyImporter] to use for fetching a users
+	// public key's for subject.
+	keyImporter PublicKeyImporter
+
+	// st is the provides the state access layer to this service.
```
[the remark this answers is not in the document]
Comment (tlm) [reply domain/keymanager/service/service.go:43]: Done & Done

**Pull request merged**